### PR TITLE
[G2M] Progress Indicator

### PR DESCRIPTION
**Pull request by Hamaad102**
Adds #132 

[Progress Indicator]
- Reworked current implementation of Process Indicator to look more in line with the design
- Added optional size prop to allow easy responsive resizing of component

**Before:**
Screenshot 2022-10-13 143333

**After:**
Screenshot 2022-10-24 132745

- Implemented vertical process bar

**Current:**
Screenshot 2022-10-24 132733

- Added optional props to provide captions for each step

**Current:**
Screenshot 2022-10-24 132721

[File Indicator]
- Modified the progress bar to serve as a file indicator
- Optional props allow for information pertaining to file size, name and time remaining for download to be displayed

**Current:**
Screenshot 2022-10-24 132816
Screenshot 2022-10-24 132834
Screenshot 2022-10-24 132846

- Also, able to display errors through optional props

**Current:**
Screenshot 2022-10-24 132901

[Step Indicator]
- Implemented the step indicator
- Added optional size prop to allow easy responsive resizing of component

**Current:**
Screenshot 2022-10-24 132658

### CODE REVIEW

**Review by vlatawiec — COMMENTED** [review]
Looks great, and interactions are great! On the horizontal/vertical version, the "bar" in between steps looks very thin, let me know if this follows the design and if so, I will update the design to increase this sizing. Thank you!

**Review by vlatawiec — APPROVED** [review]

**Comment by Hamaad102** [comment]
> Looks great, and interactions are great! On the horizontal/vertical version, the "bar" in between steps looks very thin, let me know if this follows the design and if so, I will update the design to increase this sizing. Thank you!

Yup you're right, the line is thinner. This is what it looks like now with the correct size:
Screenshot 2022-10-24 151451

File: src/components/file-indicator.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    disabledPause: {
+      pointerEvents: 'none',
+      fill: '#C8C8C8', // Secondary-400
+    },
+  })
+
+  return (
+    <div>
+      {
+        fileName && fileSize && <p className={`${style.fileName} text-secondary-700`}>{fileName} <span className={`${style.fileSize}`}>({fileSize})</span></p>
```
Comment (geoerika) [inline src/components/file-indicator.js:87]: Too long line here.

File: src/components/file-indicator.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+FileIndicator.defaultProps = {
+  classes: { root: '', content: '' },
+  animate: false,
+  direction: 'rtl',
+  duration: 10,
+  fileName: null,
+  fileSize: null,
+  progress: 100,
+  time: null,
```
Comment (geoerika) [inline src/components/file-indicator.js:143]: Add a default value for `pause` & `error` as well.

File: src/components/file-indicator.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import React, { forwardRef } from 'react'
+
+import { makeStyles, LinearAnimation } from '../utils/make-styles'
+
+const TimeConverter = (time) => {
+  const Minutes = Math.floor(time / 60)
+  const Seconds = time - Minutes * 60
+  if (Minutes) {
+    return `${Minutes} min${Minutes > 1 ? 's' : ''} and ${Seconds} second${Seconds > 1 || Seconds === 0 ? 's' : ''}`
+  } else {
```
Comment (geoerika) [inline src/components/file-indicator.js:12]: We can skip the else line here, using the logic of early return:
```
if () {
  return ...
}
return ...
```

File: src/components/progress-indicator/progress-indicator.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    </div>
+  )
+}
+
+ProgressIndicator.propTypes = {
+  indicators: PropTypes.array.isRequired,
+  classes: PropTypes.object,
+  vertical: PropTypes.bool,
+  size: PropTypes.number,
+}
```
Comment (geoerika) [inline src/components/progress-indicator/progress-indicator.js:175]: Add an empty line here.

File: src/components/progress-indicator/progress-indicator.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+          </div>
+        </div>
+      ))}
+    </div>
+  )
+}
+
+ProgressIndicator.propTypes = {
+  indicators: PropTypes.array.isRequired,
+  classes: PropTypes.object,
```
Comment (geoerika) [inline src/components/progress-indicator/progress-indicator.js:172]: We can be more specific here:
`PropTypes.objectOf(PropTypes.string)`

File: src/components/progress-indicator/progress-indicator.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+ProgressIndicator.defaultProps = {
+  classes: {
+    root: '',
+    indicatorContainer: '',
+    indicatorLabel: '',
+    indicatorLabelCaption: '',
+    indicatorNumberContainer: '',
+    indicatorNumber: '',
+  },
+}
```
Comment (geoerika) [inline src/components/progress-indicator/progress-indicator.js:185]: Add default values for all props that are not required.

File: stories/step-indicator.stories.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
+import { StepIndicator } from '../src'
+import React, { useEffect, useState } from 'react'
+
```
Comment (geoerika) [inline stories/step-indicator.stories.js:3]: Always leave 2 lines empty after imports.

File: stories/step-indicator.stories.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+const defaultIndicators = [
+  { label: 'building dashboard', active: true, complete: false },
+  { label: 'loading reports', active: false, complete: false },
+  { label: 'adding data', active: false, complete: false },
+  { label: 'testing interface', active: false, complete: false },
+  { label: 'finalizing', active: false, complete: false },
+]
+
+
```
Comment (geoerika) [inline stories/step-indicator.stories.js:25]: Delete one empty line here.

**Review by geoerika — APPROVED** [review]
Minor changes. Looks good otherwise!

File: src/components/file-indicator.js [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+  fileName: PropTypes.string,
+  fileSize: PropTypes.string,
+  progress: PropTypes.number,
+  time: PropTypes.number,
+  pause: PropTypes.func,
+  error: PropTypes.bool,
+}
+
+FileIndicator.defaultProps = {
+  classes: { root: '', content: '' },
```
Comment (kc-leung) [inline src/components/file-indicator.js:136]: we need a class selector for each of the container (at least for the main ones) so user can customize its styling. Please take a look on other component for example

File: src/components/file-indicator.js [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+const FileIndicator = forwardRef(({
+  classes, animate, direction, duration, fileName, fileSize, progress, time, pause, error = false, ...rest
+}, ref) => {
+
+  const fileIndicatorClasses = Object.freeze({ 
+    root: `h-5px w-full rounded-full bg-secondary-200 ${classes.root}`,
+    content: `h-full rounded-full bg-primary-500 ${classes.content}`,
+  })
+  const style = makeStyles({
```
Comment (kc-leung) [inline src/components/file-indicator.js:31]: lets move makestyle class outside of the component please take a look on other component as example. We should keep a format for all the components

File: src/components/progress-indicator/progress-indicator.js [hunk 11]
```diff
@@ -0,0 +1,187 @@
+import React, { useMemo } from 'react'
+import PropTypes from 'prop-types'
+import clsx from 'clsx'
+
+import { makeStyles } from '../../utils/make-styles'
+import { CheckBold } from '../../icons'
+import Spin from './spin'
```
Comment (kc-leung) [inline src/components/progress-indicator/progress-indicator.js:7]: how about adding it into icon?

File: src/components/step-indicator.js [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+        'text-primary-500': complete,
+      }, `${styles.text} capitalize`)}>{label}</p>
+    </div>
+  </div>
+)
+
+const FirstStep = (index, label, active, complete, styles) => index === 0 ? (
+  <div className={styles.container}>
+    {StepLabel({ index, label, active, complete, styles })}
+    <svg id='first' viewBox="0 0 224 38" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (kc-leung) [inline src/components/step-indicator.js:35]: how about moving svg/icon code to its own file? so it is cleaner.

**Review by kc-leung — CHANGES_REQUESTED** [review]
- Lets name progress-indicator.js to index.js since it is already in its own folder
- Are all the icons used in all theses indicators with animation and customized? If so how about creating a folder in Icons named animated since there are various type of them.
- All new components should allow user being able to target/select at least each of the important containers so the styling can be customized

File: src/components/progress-indicator/progress-indicator.js (hunk 11 above)
Comment (Hamaad102) [reply src/components/progress-indicator/progress-indicator.js]: There's a lot of unique styles for this icon that don't make it useable outside this context, it has to be white for example (different layers of opacity).

File: src/components/progress-indicator/index.js [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+      '& .label-v': {
+        marginLeft: '0.313rem',
+        fontSize: `${size/1.875}rem`,
+        letterSpacing: '0.05em',
+      },
+      '& .number-complete': {
+        width: `${size}rem`,
+        height: `${size}rem`,
+        padding: '0.313rem',
+        borderRadius: '50%',
```
Comment (woozyking) [inline src/components/progress-indicator/index.js:44]: Applying `border-radius` without resetting the border prop would produce a bit of a jagged surface (for filled/complete circles)
before 1

before 2


Add `border: 0;` for a smoother edge of a circle:
after 1

after 2

**Review by woozyking — APPROVED** [review]

File: src/components/file-indicator/index.js [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+  })
+  
+  return (
+    <div className={`filed_indicator__root-container ${style.fileIndicatorRoot}`}>
+      {
+        fileName && fileSize && 
+        <p className={`file-name font-bold mb-3 text-secondary-700 ${fileIndicatorClasses.fileName}`}>{fileName} <span className='font-normal'>({fileSize})</span></p>
+      }
+      <div className={`flex flex-row items-center gap-4 ${fileIndicatorClasses.container}`}>
+        <div ref={ref} className={fileIndicatorClasses.barContainer} {...rest}>
```
Comment (kc-leung) [inline src/components/file-indicator/index.js:92]: forwardRef should be in the root container of the component why here?

File: src/components/progress-indicator/index.js [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+    },
+  })
+}
+
+const checkComplete = (startIndex, indicators) => {
+  const rest = indicators.slice(startIndex)
+  return rest.find(({ complete }) => complete)
+}
+
+const ProgressIndicator = ({ classes, indicators: _indicators, vertical = false, size = 1.875 }) => {
```
Comment (kc-leung) [inline src/components/progress-indicator/index.js:70]: forwardRef missing

File: src/components/step-indicator/index.js [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+        fill: getTailwindConfigColor('neutral-50'),
+      },
+      '& .checkSize': {
+        width: `${size/21.33}rem`,
+      },
+    } },
+  )
+}
+
+const StepLabel = ({ index, label, active, complete }) => (
```
Comment (kc-leung) [inline src/components/step-indicator/index.js:49]: same as above

**Review by kc-leung — APPROVED** [review]
looks good lets add forwardRef for the missing ones

**Pull request merged**